src/dict.c

```diff
@@ -509,6 +510,40 @@ dictEntry *dictAddRaw(dict *d, void *key, dictEntry **existing)
     return dictInsertAtPosition(d, key, position);
 }
 
+/* Low-level add function for non-existing keys:
+ * This function adds a new entry to the dictionary, assuming the key does not
+ * already exist.
+ * Parameters:
+ * - `dict *d`: Pointer to the dictionary structure.
+ * - `void *key`: Pointer to the key being added.
+ * - `const uint64_t *hash`: Optional hash of the key being added. If not NULL will avoid
+ *  recalculating it.
+ * Guarantees:
+ * - The key is assumed to be non-existing, and its corresponding bucket is found
+ *   using `dictFindInsertForNonExistingKey`.
+ * - The function will assert if `dictFindInsertForNonExistingKey` returns NULL,
+ *   which should not happen under correct usage.
+ * Note:
+ * Ensure that the key's uniqueness is managed externally before calling this function. */
+dictEntry *dictAddNonExistingRaw(dict *d, void *key, const uint64_t *hash)
+{
+    uint64_t h;
+    /* Calculate the hash only if it's not provided */
+    if (hash == NULL) {
+        h = dictHashKey(d, key, d->useStoredKeyApi);
+    } else {
+        h = *hash;
+    }
+    /* Get the position for the new key, it should never be NULL. */
+    void *position = dictFindInsertForNonExistingKey(d, h);
+    assert(position!=NULL);
+
+    /* Dup the key if necessary. */
+    if (d->type->keyDup) key = d->type->keyDup(d, key);
+
+    return dictInsertAtPosition(d, key, position);
+}
```

Comment:
I don't think there is real use to pass hash as NULL. The only way we can indicate that the key is not exists in the dict is only after we tried to search for it - and in that case we should have its hash available:

```suggestion
dictEntry *dictAddNonExistingRaw(dict *d, void *key, const uint64_t hash) {
    /* Get the position for the new key, it should never be NULL. */
    void *position = dictFindInsertForNonExistingKey(d, hash);
    assert(position!=NULL);

    /* Dup the key if necessary. */
    if (d->type->keyDup) key = d->type->keyDup(d, key);

    return dictInsertAtPosition(d, key, position);
}
```